File: mobile/android/OpenU60/app/src/main/java/com/openu60/feature/router/networkmode/NetworkModeViewModel.kt
package com.openu60.feature.router.networkmode

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openu60.core.model.NetworkModeConfig
import com.openu60.core.model.NetworkModeParser
import com.openu60.core.network.AgentClient
import com.openu60.core.network.AgentError
import com.openu60.core.network.AuthManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class NetworkModeState(
    val config: NetworkModeConfig = NetworkModeConfig.empty,
    val isLoading: Boolean = false,
    val message: String? = null,
    val messageIsError: Boolean = false,
)

@HiltViewModel
class NetworkModeViewModel @Inject constructor(
    private val agentClient: AgentClient,
    private val authManager: AuthManager,
) : ViewModel() {

    private val _state = MutableStateFlow(NetworkModeState())
    val state: StateFlow<NetworkModeState> = _state.asStateFlow()

    fun refresh() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val data = agentClient.getJSON("/api/modem/network-mode")
                val config = NetworkModeParser.parse(data)
                _state.value = _state.value.copy(config = config, isLoading = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) refresh() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun setNetworkMode(value: String) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.putJSON("/api/modem/network-mode", mapOf("net_select" to value))
                repeat(5) {
                    delay(2000)
                    try {
                        val data = agentClient.getJSON("/api/modem/network-mode")
                        val config = NetworkModeParser.parse(data)
                        if (config.netSelect == value) {
                            _state.value = _state.value.copy(
                                config = config,
                                isLoading = false,
                                message = "Network mode updated",
                                messageIsError = false,
                            )
                            return@launch
                        }
                    } catch (_: Exception) {}
                }
                _state.value = _state.value.copy(
                    config = _state.value.config.copy(netSelect = value),
                    isLoading = false,
                    message = "Mode change sent",
                    messageIsError = false,
                )
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) setNetworkMode(value) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    private fun setError(msg: String?) {
        _state.value = _state.value.copy(isLoading = false, message = msg ?: "Unknown error", messageIsError = true)
    }
}

File: mobile/ios/OpenU60/Features/RouterSettings/NetworkMode/NetworkModeViewModel.swift
import SwiftUI

@Observable
@MainActor
final class NetworkModeViewModel {
    var config: NetworkModeConfig = .empty
    var isLoading: Bool = false
    var message: String?
    var messageIsError: Bool = false

    var selectedNetSelect: String = NetworkModeConfig.netSelectOptions[0].value

    private let client: AgentClient
    private let authManager: AuthManager

    init(client: AgentClient, authManager: AuthManager) {
        self.client = client
        self.authManager = authManager
    }

    func refresh() async {
        isLoading = true
        message = nil

        do {
            let data = try await client.getJSON("/api/network/signal")
            config = NetworkModeParser.parse(data)
            selectedNetSelect = config.netSelect
        } catch {
            showMessage("Failed to load network mode: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func applyMode() async {
        isLoading = true

        do {
            if selectedNetSelect != config.netSelect {
                let _ = try await client.putJSON("/api/modem/network-mode", body: ["net_select": selectedNetSelect])
            }
            // Poll until the router confirms the new value (up to ~10s)
            let expectedNet = selectedNetSelect
            for _ in 0..<5 {
                try? await Task.sleep(for: .seconds(2))
                let data = try await client.getJSON("/api/network/signal")
                let fetched = NetworkModeParser.parse(data)
                if fetched.netSelect == expectedNet {
                    config = fetched
                    showMessage("Network mode updated", isError: false)
                    isLoading = false
                    return
                }
            }

            config = NetworkModeConfig(netSelect: expectedNet)
            showMessage("Mode sent — router may still be switching", isError: false)
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    private func showMessage(_ text: String, isError: Bool) {
        message = text
        messageIsError = isError
    }
}
